### DISCUSSION TITLE
Deploying FastAPI as Lambda Function

### DISCUSSION THREAD

**Discussion by chgangaraju** [Kludex/mangum#221] [Q&A]
from fastapi import FastAPI
    from mangum import Mangum

    app = FastAPI()

    @app.get("/")
    def hello_world():
        return {"hello": "world"}

    handler = Mangum(app)

ERROR:

    [ERROR] TypeError: __call__() takes 2 positional arguments but 4 were given
    Traceback (most recent call last):
      File "/var/task/mangum/__init__.py", line 123, in __call__
        raise exc
      File "/var/task/mangum/__init__.py", line 118, in __call__
        response = self.asgi(*args, **kwargs)
      File "/var/task/mangum/__init__.py", line 178, in asgi
        response = asgi_cycle(self.app, body=body)
      File "/var/task/mangum/__init__.py", line 38, in __call__
        asgi_instance = app(self.scope, self.receive, self.send)

I tried this, but could not import  `AWSLambdaAdapter`

    from mangum.platforms.aws.adapter import AWSLambdaAdapter

Can you please provide a working example of this?

**Comment by chgangaraju** [comment]
#22

**Comment by jordaneremieff** [comment]
@chgangaraju the previous `AWSLambdaAdapter` class was renamed `Mangum`, and the default spec version is ASGI 3. In order to use FastAPI, you will need to set the `spec_version` to 2:

```python
from fastapi import FastAPI
from mangum import Mangum

app = FastAPI()

@app.get("/")
def hello_world():
    return {"hello": "world"}

handler = Mangum(app, spec_version=2)
```

**Comment by chgangaraju** [comment]
Thanks, @erm. It is working like a charm. It will be very helpful if this is updated in docs and [examples](https://github.com/erm/asgi-examples/tree/master/fastapi).

**Comment by aniljaiswal** [comment]
@chgangaraju Can you share your serverless.yml function setup? I want to know how you have organized the routing for FastAPI to handle and how does your project structure look like. I'm having a hard time figuring out doing it.

**Comment by chgangaraju** [comment]
serverless.yml


    functions:
      app:
        handler: app.main.handler
        events:
          - http: ANY /
          - http:
              path: /{proxy+}
              method: ANY

app/main.py

```
import uvicorn
from fastapi import FastAPI
from mangum import Mangum

app = FastAPI()
handler = Mangum(app)


@app.get("/")
def hello_world():
    return {"hello": "world"}


# To run locally 
if __name__ == '__main__':
    uvicorn.run(app)
```

Make sure that you are using a compatible python version. For example, If you are using Python 3.6, use mangum==0.6.6. As the latest version only works with Python 3.7+.

**Comment by jordaneremieff** [comment]
@chgangaraju what are the reasons you are using Python 3.6/Mangum 0.6.6? You brought up 3.6 support in a previous issue https://github.com/erm/mangum/issues/33, and I'm still open to a potential refactor to support 3.6 again if there are specific reasons to do so.

Also, FastAPI currently supports for ASGI 3, so I dropped `spec_version` at some point as well.

**Comment by chgangaraju** [comment]
@erm  We have multiple projects and we are using Python 3.6 and flask, connexion in almost all projects. We experimented to use FastAPI& Mangum for one lambda.  Even though it's easy to switch python versions with tools like `pyenv`. it's a little bit difficult to change CI/CD pipelines for a specific project. We will slowly switch to 3.7 soon.

**Comment by chgangaraju** [comment]
@anil-grexit, Refer to this working [example](https://github.com/chgangaraju/fastapi-mangum-example)

@erm, It's very hard to find a working example with the latest versions. I tried [this example](https://github.com/erm/mangum-examples), but getting the following error.

    [ERROR] TypeError: __call__() missing 2 required positional arguments: 'receive' and 'send'
    Traceback (most recent call last):
      File "/var/task/mangum/__init__.py", line 93, in __call__
        raise exc
      File "/var/task/mangum/__init__.py", line 88, in __call__
        response = self.asgi(*args, **kwargs)
      File "/var/task/mangum/__init__.py", line 140, in asgi
        response = ASGICycle(scope, binary=binary)(self.app, body=body)
       File "/var/task/mangum/__init__.py", line 27, in __call__
        asgi_instance = app(self.scope)

**Comment by jordaneremieff** [comment]
@chgangaraju The examples in that repo may be out-of-date (will update or archive it at some point), but I don't see anything obviously wrong with the application code that would produce that error with the latest versions.

The main difficulty around creating examples has been that there are several different configuration/deployment methods which I am now treating as a separate concern from the adapter class itself. Initially I bundled a custom CLI for generating CF templates that I've since pulled out so some of the existing example code may not reflect this yet.

Thanks for putting together your serverless repo. I'll take a look into what issues there may be between the latest version of the adapter against your example when I have a chance.

**Comment by jordaneremieff** [comment]
@chgangaraju I've tested against your example repo. If the requirements are changed to use the latest versions of FastAPI and Mangum with Python 3.7 there are no issues. I think there may be problems when using versions of FastAPI that do not support ASGI 3 that are causing the error above, but I was unable to reproduce your specific error.

**Comment by aroraprince** [comment]
I think the issue still persists while deploying FastAPI, using Mangum, and Serverless Framework.

I have tried a couple of things till I am not able to get the whole setup working using the latest library versions:

My handler:
```
from fastapi import FastAPI
from mangum import Mangum


app = FastAPI()
handler = Mangum(app)

@app.get("/ping")
def ping():
    return {'response': 'pong'}

```

serverless.yml:
```
provider:
  name: aws
  runtime: python3.8
  stage: dev
  region: ap-southeast-1
  memorySize: 256

functions:
  ping:
    handler: ping.handler
    events:
      - http:
          path: ping
          method: get
          cors: true
```


My requirements.txt

```
appnope==0.1.0
backcall==0.2.0
certifi==2020.6.20
chardet==3.0.4
click==7.1.2
decorator==4.4.2
fastapi==0.61.1
h11==0.9.0
httpcore==0.10.2
httptools==0.1.1
httpx==0.14.1
idna==2.10
ipython==7.17.0
ipython-genutils==0.2.0
jedi==0.17.2
mangum==0.9.2
parso==0.7.1
pexpect==4.8.0
pickleshare==0.7.5
prompt-toolkit==3.0.6
ptyprocess==0.6.0
pydantic==1.6.1
Pygments==2.6.1
requests==2.24.0
rfc3986==1.4.0
six==1.15.0
sniffio==1.1.0
starlette==0.13.6
traitlets==4.3.3
typing-extensions==3.7.4.2
urllib3==1.25.10
uvicorn==0.11.8
uvloop==0.14.0
wcwidth==0.2.5
websockets==8.1
```


When I deploy it using serverless, sls depoy and Invoke the function I ma getting the following error:

```
[ERROR] KeyError: 'requestContext'
Traceback (most recent call last):
  File "/var/task/mangum/adapter.py", line 110, in __call__
    return self.handler(event, context)
  File "/var/task/mangum/adapter.py", line 130, in handler
    if "eventType" in event["requestContext"]:
```

I have tried with python 3.8 and 3.7.
Not able to find a resolution on the web for the same.
Also tried using the parameters spec_version=2(which is not required I feel).

I feel something is missing here.

**Comment by jordaneremieff** [comment]
@aroraprince Hello. There is another issue that I think is related to the issue you're describing here https://github.com/jordaneremieff/mangum/issues/126. The ASGI connection scope is built using information provided in the request context when invoked by API Gateway, it appears you are invoking the function without the request context here.

**Comment by aroraprince** [comment]
@jordaneremieff yes, I think it is not able to pass the request context.
As mentioned in #126 , I will check on StackOverflow once again, and post the solution online.

**Comment by TAJD** [comment]
A quick note for anyone who is fresh to deploying FastAPI via AWS API Gateways as a Proxy resource - you must format your response in the following manner:

```python
    msg = useful_function(accept_arguments)
    headers = { # example headers
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'OPTIONS,POST,GET'
    }
    return {
    'statusCode': 200,
    'headers': headers,
    'body': json.dumps(msg)
    }
```

Otherwise, you can have all sorts of fun dealing with CORS errors having otherwise correctly configured everything else within FastAPI and AWS.

**Comment by kmcquade** [comment]
For anyone else who runs into this issue, just to elaborate on what @TAJD said... 

Using the `CORSMiddleware` class and `add_middleware` accomplishes the same that he mentioned in his comment. This is shown in this example:
 
https://github.com/chgangaraju/fastapi-mangum-example/blob/master/app/main.py


And shown below for your convenience, with comments.

```python
#!/usr/bin/env python3

import uvicorn
from fastapi import FastAPI
from mangum import Mangum
# TODO: Add this line
from starlette.middleware.cors import CORSMiddleware

app = FastAPI(title="FastAPI Mangum Example", version='1.0.0')

# TODO: Add these lines
app.add_middleware(
    CORSMiddleware,
    allow_origins='*',
    allow_credentials=False,
    allow_methods=["GET", "POST", "OPTIONS"],
    allow_headers=["x-apigateway-header", "Content-Type", "X-Amz-Date"],
)

handler = Mangum(app)


@app.get('/', name='Hello World', tags=['Hello'])
def hello_world():
    return {"Hello": "Python"}


if __name__ == '__main__':
    uvicorn.run(app)
```

**Comment by federicodv** [comment]
> I think the issue still persists while deploying FastAPI, using Mangum, and Serverless Framework.
> 
> I have tried a couple of things till I am not able to get the whole setup working using the latest library versions:
> 
> My handler:
> 
> ```
> from fastapi import FastAPI
> from mangum import Mangum
> 
> 
> app = FastAPI()
> handler = Mangum(app)
> 
> @app.get("/ping")
> def ping():
>     return {'response': 'pong'}
> ```
> 
> serverless.yml:
> 
> ```
> provider:
>   name: aws
>   runtime: python3.8
>   stage: dev
>   region: ap-southeast-1
>   memorySize: 256
> 
> functions:
>   ping:
>     handler: ping.handler
>     events:
>       - http:
>           path: ping
>           method: get
>           cors: true
> ```
> 
> My requirements.txt
> 
> ```
> appnope==0.1.0
> backcall==0.2.0
> certifi==2020.6.20
> chardet==3.0.4
> click==7.1.2
> decorator==4.4.2
> fastapi==0.61.1
> h11==0.9.0
> httpcore==0.10.2
> httptools==0.1.1
> httpx==0.14.1
> idna==2.10
> ipython==7.17.0
> ipython-genutils==0.2.0
> jedi==0.17.2
> mangum==0.9.2
> parso==0.7.1
> pexpect==4.8.0
> pickleshare==0.7.5
> prompt-toolkit==3.0.6
> ptyprocess==0.6.0
> pydantic==1.6.1
> Pygments==2.6.1
> requests==2.24.0
> rfc3986==1.4.0
> six==1.15.0
> sniffio==1.1.0
> starlette==0.13.6
> traitlets==4.3.3
> typing-extensions==3.7.4.2
> urllib3==1.25.10
> uvicorn==0.11.8
> uvloop==0.14.0
> wcwidth==0.2.5
> websockets==8.1
> ```
> 
> When I deploy it using serverless, sls depoy and Invoke the function I ma getting the following error:
> 
> ```
> [ERROR] KeyError: 'requestContext'
> Traceback (most recent call last):
>   File "/var/task/mangum/adapter.py", line 110, in __call__
>     return self.handler(event, context)
>   File "/var/task/mangum/adapter.py", line 130, in handler
>     if "eventType" in event["requestContext"]:
> ```
> 
> I have tried with python 3.8 and 3.7.
> Not able to find a resolution on the web for the same.
> Also tried using the parameters spec_version=2(which is not required I feel).
> 
> I feel something is missing here.

i'm having same issue and is impossible to find a working example with an updated version
i.e: 
serverless invoke local --function app 
```
Traceback (most recent call last):
  File "/home/federico/.serverless/artifacts/2.38.0/runtimeWrappers/invoke.py", line 86, in <module>

    result = handler(input['event'], context)
  File "..../venv/lib/python3.8/site-packages/mangum/adapter.py", line 101, in __call__
    scope = self.create_scope(event, context)
  File "..../venv/lib/python3.8/site-packages/mangum/adapter.py", line 118, in create_scope
    request_context = event["requestContext"]
KeyError: 'requestContext'

**Comment by saharavishag** [comment]
Experiencing the same issue :( @federicodv Did you manage to resolve it somehow?

**Reply by himalacharya** [reply]
Still this issue. Downgrading mangum did not work

**Reply by jordaneremieff** [reply]
@saharavishag @himalacharya my comment [here](https://github.com/jordaneremieff/mangum/discussions/218#discussioncomment-2071335) also applies to this discussion. The local invocations are not passing an event payload that includes the `requestContext`, so the handler is unable to build the `scope` required to run the request/response cycle.

**Reply by himalacharya** [reply]
@jordaneremieff  My app works all time except calling scheduled event. When I cancel the scheduled event, no error generates and code works fine. 
My even scheduler runs every 5 min and defined as 
`{   "isBase64Encoded": false,   "statusCode": 200,   "body": "{ \"message\": \"Hello from Lambda!\" }",   "headers": {     "content-type": "application/json"   } }`
Using FastApi,AWS lambda and API gateway. 
```
[ERROR] KeyError: 'requestContext'
Traceback (most recent call last):
  File "/opt/python/lib/python3.8/site-packages/codeguru_profiler_agent/aws_lambda/profiler_decorator.py", line 52, in profiler_decorate
    return function(event, context)
  File "/opt/python/lib/python3.8/site-packages/codeguru_profiler_agent/aws_lambda/lambda_handler.py", line 91, in call_handler
    return handler_function(event, context)
  File "/var/task/mangum/adapter.py", line 101, in __call__
    scope = self.create_scope(event, context)
  File "/var/task/mangum/adapter.py", line 118, in create_scope
    request_context = event["requestContext"]
```

**Reply by jordaneremieff** [reply]
Yes, this is expected. The event does not contain the `requestContext` key which is sent by API Gateway.

**Reply by himalacharya** [reply]
How can we solve the issue for scheduling event every 5 minutes just to warm Lambda? I just need to call lambda every 5 minutes. I created a rule in Cloudwatch and configure input as constant(json text) as { "isBase64Encoded": false, "statusCode": 200, "body": "{ \"message\": \"Hello from Lambda!\" }", "headers": { "content-type": "application/json" } }

**Reply by aniljaiswal** [reply]
@saharavishag Why not use `Provisioned concurrency` to take care of the cold start issue?

> Provisioned concurrency – Provisioned concurrency initializes a requested number of execution environments so that they are prepared to respond immediately to your function's invocations. Note that configuring provisioned concurrency incurs charges to your AWS account.

Source: https://docs.aws.amazon.com/lambda/latest/dg/configuration-concurrency.html

**Reply by AlphonsG** [reply]
> Experiencing the same issue :( @federicodv Did you manage to resolve it somehow?

run this when testing locally:

 curl -XPOST "http://localhost:8080/2015-03-31/functions/function/invocations/predict" -d '{"resource": "/", "path": "/api/v1/test/", "httpMethod": "GET", "requestContext": {}, "multiValueQueryStringParameters": null}'
